### DISCUSSION TITLE
Business Logic

### DISCUSSION THREAD

**Discussion by buuum** [directus/directus#6776] [Archive]
Hello,

my user model has credits. For the user to be able to publish a post he must have more than 5 credits. should I put this business logic in the "create.before" hooks?

Thanks

**Comment by philleepflorence** [comment]
### Follow Up Questions
- How are the credits managed?
- How is `publish` handled - during item `create`, `update` or via field `status`?
- Have you looked into [Roles - API Reference](https://docs.directus.io/reference/api/system/roles/)
    - such as upgrading to a role based on certain criteria.

**Reply by buuum** [reply]
The user buys credits and they are added. (not implemented yet)
When the user creates a post he has to have enough credits to do it and at the same time they will be discounted once created.
This part I don't see how to do it with roles.

**Reply by philleepflorence** [reply]
EDIT: Roles in Directus 9 are quite granular and extensive:  
> Roles define a specific set of access permissions, and are the primary organizational structure for Users within the platform.
> See [Roles - Concept Reference](https://docs.directus.io/concepts/roles/)
- You can always perform runtime logic computation using the API Event Hooks, 
    - but roles manage item layout, access, and edit privileges more efficiently.
- If _Creating a post_ === _Publishing a post_, and 5 credits create a post - `discounted === subtracted`
    - Consider upgrading the user role when `credits >= 5` and downgrading when `credits <= 5` using the `items.update.after` Event Hook.

**Comment by benhaynes** [comment]
Combining `create.before` and `create.after` would make sense to me. Either add a `credits` collection and track it relationally as `directus_users.credits` (O2M) ... or just add a static `directus_users.credits` (INT). Then:

1. `create.before` — Look up user credits using the `$CURRENT_USER`
2. `create.before` — If there are more than 5 credits, allow action
3. `create.after` — After _successful_ creation, remove 5 credits from `$CURRENT_USER`

This has the added benefit of only removing the credit if the **create** is successful.

**Reply by philleepflorence** [reply]
I should also add that: 
> the User might be able to `submit` a new item (esp. in the App), but the Event Hook might reject the `creation`; 
> as opposed to using Event Hook to update the Role and Credits of the User - preventing access and edit process vs intercepting submission event.

I might also be biased towards Data State Hooks (SQL Triggers) vs Request Runtime Hooks.
But, I suppose the implementation could be similar to a Simplified Subscriptions Model?

**Reply by buuum** [reply]
I find it interesting credit collection.
By having a credits collection I can add the stripe logic its hook when creating right?
